internal/controller/nbresource_controller.go

```diff
@@ -109,82 +109,108 @@ func (r *NBResourceReconciler) handlePolicy(ctx context.Context, req ctrl.Reques
-			nbPolicy.Status.LastUpdatedAt = &v1.Time{Time: time.Now()}
-			updatePolicyStatus = true
-		}
+			err := r.Client.Get(ctx, types.NamespacedName{Name: policy}, &nbPolicy)
+			if err != nil {
+				logger.Error(errKubernetesAPI, "error getting NBPolicy", "err", err, "policyName", policy)
+				return err
+			}
+			if !util.Contains(statusPolicies, policy) {
+				// New
+				if !util.Contains(nbPolicy.Status.ManagedServiceList, req.NamespacedName.String()) {
+					nbPolicy.Status.ManagedServiceList = append(nbPolicy.Status.ManagedServiceList, req.NamespacedName.String())
+					updatePolicyStatus = true
+				}
+			} else {
+				// Check update
+				if !util.Contains(nbPolicy.Status.ManagedServiceList, req.NamespacedName.String()) {
+					nbPolicy.Status.ManagedServiceList = append(nbPolicy.Status.ManagedServiceList, req.NamespacedName.String())
+					nbPolicy.Status.LastUpdatedAt = &v1.Time{Time: time.Now()}
+					updatePolicyStatus = true
+				}
 
-		if !util.Equivalent(nbResource.Spec.TCPPorts, nbResource.Status.TCPPorts) {
-			nbResource.Status.TCPPorts = nbResource.Spec.TCPPorts
-			nbPolicy.Status.LastUpdatedAt = &v1.Time{Time: time.Now()}
-			updatePolicyStatus = true
-		}
+				if !util.Equivalent(nbResource.Spec.TCPPorts, nbResource.Status.TCPPorts) {
+					nbResource.Status.TCPPorts = nbResource.Spec.TCPPorts
+					nbPolicy.Status.LastUpdatedAt = &v1.Time{Time: time.Now()}
+					updatePolicyStatus = true
+				}
+
+				if !util.Equivalent(nbResource.Spec.UDPPorts, nbResource.Status.UDPPorts) {
+					nbResource.Status.UDPPorts = nbResource.Spec.UDPPorts
+					nbPolicy.Status.LastUpdatedAt = &v1.Time{Time: time.Now()}
+					updatePolicyStatus = true
+				}
+
+				if !util.Equivalent(nbResource.Status.Groups, groupIDs) {
+					nbResource.Status.Groups = groupIDs
+					nbPolicy.Status.LastUpdatedAt = &v1.Time{Time: time.Now()}
+					updatePolicyStatus = true
+				}
+			}
 
-		if !util.Equivalent(nbResource.Spec.UDPPorts, nbResource.Status.UDPPorts) {
-			nbResource.Status.UDPPorts = nbResource.Spec.UDPPorts
-			nbPolicy.Status.LastUpdatedAt = &v1.Time{Time: time.Now()}
-			updatePolicyStatus = true
+			if updatePolicyStatus {
+				err := r.Client.Status().Update(ctx, &nbPolicy)
+				if err != nil {
+					logger.Error(errKubernetesAPI, "error updating NBPolicy", "err", err, "policyName", policy)
+					return err
+				}
+			}
 		}
 
-		if !util.Equivalent(nbResource.Status.Groups, groupIDs) {
-			nbResource.Status.Groups = groupIDs
-			nbPolicy.Status.LastUpdatedAt = &v1.Time{Time: time.Now()}
-			updatePolicyStatus = true
+		for _, policy := range statusPolicies {
+			// Delete
+			if !util.Contains(specPolicies, policy) {
+				err := r.Client.Get(ctx, types.NamespacedName{Name: policy}, &nbPolicy)
+				if !errors.IsNotFound(err) {
+					if err != nil {
+						logger.Error(errKubernetesAPI, "error getting NBPolicy", "err", err, "policyName", policy)
+						return err
```

Comment:
why do we return err here, isn't that also the desired state?

Reply:
It's not a notFound error, which means it could be a temporary network/control plane error, so it could be an NBPolicy which needs to be updated and may/may not exist but we can't know since control plane is errored in this case